src/java.base/share/classes/java/util/BitSet.java

```diff
@@ -384,9 +415,11 @@ public void flip(int bitIndex) {
 
         int wordIndex = wordIndex(bitIndex);
         expandTo(wordIndex);
-
-        words[wordIndex] ^= (1L << bitIndex);
-
+        
+        final long bitMask = 1L << bitIndex;
+        words[wordIndex] ^= bitMask;
+        cardinality += (words[wordIndex] & bitMask) != 0 ? 1 : -1;
```

Comment:
Isn't there a race condition here? 

I'm wondering if it would not be better to cache the last cardinality and invalidate (negative cardinality) when the bit set is changed. Less intrusive and potentially quicker.

Reply:
A perhaps slightly less racy way would be to only read and write from `words` once:
```
final long newValue = words[wordIndex] ^ bitMask;
cardinality += (newValue & bitMask) != 0 ? 1 : -1;
words[wordIndex] = newValue;
```
.. but the fact remains that `BitSet` is not thread-safe, so I think we shouldn't complicate things to avoid or eliminate potential races. (Using a local might help the compiler to avoid the back-to-back array read, but probably doesn't matter)

Reply:
I think the same of @cl4es.